### PR TITLE
Handle single primer-trimming

### PR DESCRIPTION
**Pull request by FelixMoelder**
These are first adjustment of the pipeline when receiving just one primer for trimming.
In this case primers can not be used for region targeting and a separat bed-file is required.
As primers/roi-files often hold coordinates of hg19 a liftover chain url can be defined allowing to perform a liftover for targeted regions.

This will be finalized as soon as fgbio is capable of handling single primers.

### CODE REVIEW

File: workflow/rules/common.smk
Comment (dlaehnemann) [inline workflow/rules/common.smk:243]: If possible, I'd specify the genome version of the `"target_regions"` file here and then automatically download the correct liftover file based on this genome versions and the genome version used for mapping in the workflow.

**Review by dlaehnemann — CHANGES_REQUESTED** [review]
This generally looks good, only one bit where I would prefer a bit more of an automagic behaviour... :mage:

File: workflow/rules/common.smk
Comment (FelixMoelder) [reply workflow/rules/common.smk:243]: Should be possible as long the Syntax would be like `<source>To<target>` as that seems to the the syntax which is used by [ucsc](https://hgdownload.soe.ucsc.edu/goldenPath/hg19/liftOver/)

Comment (dlaehnemann) [reply workflow/rules/common.smk:243]: Ah right, they have their own genome version naming scheme. So this will not go together with our genome version specification. So either we map genome version specifications from one to the other, but would also have to do this for other species, or we do one of these:

1. The user does have to specify the liftover file, like you suggested.
2. We find another smart way of determining the regions of interest automatically. Maybe something that uses `mosdepth` to get the regions of the genome that are covered, and of those select only the regions that overlap with at least one of the primers?

Comment (dlaehnemann) [reply workflow/rules/common.smk]: Yeah, it's a good start. But the danger is that it goes out of sync with the general genome specified for the workflow. Maybe something to discuss with Johannes at some point. But we can leave it like this, for now, and touch it again once we have fgbio working for single primers.

**Comment by FelixMoelder** [comment]
This won't run until single primer handling is included in fgbio (see https://github.com/fulcrumgenomics/fgbio/pull/681).

**Comment by FelixMoelder** [comment]
The last commit changed the order of some rules in the workflow.
Previously, it was not possible to enable consensus read calculation and primer trimming as only capable of trimming paired end data which does not exist after calculation of consensus reads.
This is now solved by performing primer trimming first.
Moreover the workflow preferred consensus calculation over duplicate marking/deduplication.
As marking duplicates can be applied as a preprocessing step for merging duplicates into a single read both rules will now be applied if enabled.

Base recalibration will now be applied after trimming, deduplication and consensus calculation.

File: workflow/envs/filter_reads.yaml
Comment (mbargull) [inline workflow/envs/filter_reads.yaml:5]: maybe try
```suggestion
  - rust
  - compiler-c
```
here to get a `${CC}` which `bzip2-sys`'s build can use.

File: workflow/scripts/filter_primers.rs
Comment (tedil) [inline workflow/scripts/filter_primers.rs:1]:
```suggestion
```
you don't actually need the shebang when using rust-scripts in snakemake contexts.

**Comment by FelixMoelder** [comment]
This PR is ready now but we still need to wait for a [new release](https://github.com/fulcrumgenomics/fgbio/issues/715) of fgbio before merging.

File: workflow/envs/filter_reads.yaml
Comment (johanneskoester) [inline workflow/envs/filter_reads.yaml:9]: Please also pin the other versions here.

File: workflow/envs/bowtie.yaml
Comment (johanneskoester) [inline workflow/envs/bowtie.yaml:5]: Please pin bowtie.

File: workflow/rules/calling.smk
Comment (johanneskoester) [inline workflow/rules/calling.smk:41]: No need for the function anymore?

**Review by johanneskoester — CHANGES_REQUESTED** [review]
Epic work! Please find my comments below.

File: workflow/rules/calling.smk
Comment (FelixMoelder) [reply workflow/rules/calling.smk:41]: Actually no. This had to be removed due to restructuring.
Previously the input bams for `varlociraptor_preprocess` were either the recalibrated bams or alternatively output of primer trimming. So, recalibration happened before trimming.
As there were some incompatibilities between newly introduced steps base recalibration has moved to the end of the preprocessing workflow so the input needed by `varlociraptor_preprocess` will always be the recalibrated one.

File: workflow/Snakefile
Comment (FelixMoelder) [reply workflow/Snakefile:48]: Nope, this was just left behind from testing.

File: config/config.yaml
Comment (johanneskoester) [reply config/config.yaml:26]: Ah, and then in the sample sheet you define the panel? Is this already documented?

Comment (johanneskoester) [reply config/config.yaml:26]: THis should be explained in depth in config/README.md

Comment (FelixMoelder) [reply config/config.yaml:26]: Good point. This appeared intuitive to me but it is not.

File: config/README.md
Comment (johanneskoester) [inline config/README.md:33]:
```suggestion
When primer trimming is enabled, primers have to be defined either directly in the `config.yaml` or in a seperate tsv-file.
```

File: config/README.md
Comment (johanneskoester) [inline config/README.md:38]:
```suggestion
For single primer trimming only, the first entry in the config (respective in the tsv file) needs to be defined.
```

File: config/README.md
Comment (johanneskoester) [inline config/README.md:31]: Add an introductory sentence mentioning that the pipeline allows to trim amplicon primers (both ends of a fragment and only one single end of a fragment). Specify whether the latter is supposed to be the left or the right end.

**Review by johanneskoester — CHANGES_REQUESTED** [review]

File: config/README.md
Comment (johanneskoester) [inline config/README.md:34]: supposed

File: config/README.md
Comment (johanneskoester) [inline config/README.md:35]: enabled,

**Review by johanneskoester — CHANGES_REQUESTED** [review]

**Review by johanneskoester — APPROVED** [review]

**Pull request merged**
